### PR TITLE
hf dict cfg overrides

### PR DESCRIPTION
**Pull request by vchiley**
fixes issue: https://github.com/mosaicml/llm-foundry/issues/89

### CODE REVIEW

**Review by dakinggg — APPROVED** [review]
Maybe add a unit test?

File: llmfoundry/models/hf/hf_causal_lm.py [hunk 1]
```diff
@@ -47,7 +47,19 @@ def __init__(self, om_model_config: DictConfig, tokenizer: Tokenizer):
...
+        )
+
+        # set config overrides
+        for k, v in om_model_config.get('config_overrides', {}).items():
+            if not hasattr(config, k):
+                raise ValueError(
+                    f'config does not have attribute "{k}" to override ({k}: {v}).'
+                )
+
+            if isinstance(getattr(config, k), Dict):
```
Comment (dakinggg) [inline llmfoundry/models/hf/hf_causal_lm.py:59]:
```suggestion
            if isinstance(getattr(config, k), Mapping):
```

might as well be permissive

File: llmfoundry/models/hf/hf_causal_lm.py [hunk 2]
```diff
@@ -47,7 +47,19 @@ def __init__(self, om_model_config: DictConfig, tokenizer: Tokenizer):
...
+                getattr(config, k).update(v)
```
Comment (dakinggg) [inline llmfoundry/models/hf/hf_causal_lm.py:60]: do you want to check that any keys in `v` are present in `k`? that would prevent typos in the nested config from getting silently ignored.

Comment (vchiley) [reply llmfoundry/models/hf/hf_causal_lm.py:60]: I was going to, but then realized that init_config might args which are not part of the default setting so i decided to let this checking slide and hope for the best... (pulling some HF ish)

Comment (dakinggg) [reply llmfoundry/models/hf/hf_causal_lm.py:60]: is it worth defaulting all args in the init defaults?

**Comment by alextrott16** [comment]
Should these same changes be incorporated into `hf_prefix_lm.py` and `hf_t5.py`?

File: llmfoundry/models/hf/hf_causal_lm.py (hunk 2 above)
Comment (vchiley) [reply llmfoundry/models/hf/hf_causal_lm.py:66]: sure

File: llmfoundry/models/hf/hf_causal_lm.py [hunk 3]
```diff
@@ -47,7 +47,25 @@ def __init__(self, om_model_config: DictConfig, tokenizer: Tokenizer):
...
+                    f'config does not have attribute "{k}" to override ({k}: {v}).'
+                )
+
+            attr = getattr(config, k)
+            if isinstance(attr, Mapping):
+                extra_keys = [_k for _k in v.keys() if _k not in attr.keys()]
+                if extra_keys:
+                    raise ValueError(f'Config dict override got unknown keys. '
+                                     f'Extra keys: {extra_keys}. '
+                                     f'Expected keys: {list(attr.keys())}.')
```
Comment (dakinggg) [inline llmfoundry/models/hf/hf_causal_lm.py:65]: "Expected (a subset of) keys: ...`

**Review by dakinggg — APPROVED** [review]
LGTM

**Pull request merged**